compiler/ir/dart/scheduler.py

```diff
@@ -1,43 +1,90 @@
+from collections.abc import Generator
+
 from compiler.ir.dart.access_pattern import Schedule, Template
 
 
-def scheduler(template: Template, schedule: Schedule) -> Schedule:
-    for i in range(template.num_dims):
-        # i = 0: look at the last dimension
-        # i = 1: look at the second to last dimension
-        template_dim = template.num_dims - i - 1
-        schedule_dim = schedule.num_dims - i - 1
-        match = False
-
-        # maximum number of rotations
-        for _ in range(schedule_dim + 1):
-            # check if there is a match
-            template_check = template.disable_dims(template_dim)
-            schedule_check = schedule.disable_dims(schedule_dim)
-
-            if template_check.matches(schedule_check):
-                match = True
-                break
-
-            # else rotate the for loops
-            schedule = schedule.rotate(schedule_dim + 1)
-
-        if not match:
-            raise RuntimeError("failed to match template and schedule")
-
-        # now, check bounds and design potential transformation map
-        if not (template_bound := template[0].bounds[template_dim]):
-            # nothing to worry about, continue to next dim
+def scheduler_backtrack(
+    template: Template,
+    schedule: Schedule,
+    inner_dims: int = 1,
+) -> Generator[Schedule]:
+    """
+    Backtracking method to find all possible mappings of the schedule on the template
+
+    `template` (Template): the accelerator template
+    `schedule` (Schedule): the partially scheduled operation
+    `inner_dims` (int): current number of innermost dimensions being handled
+    `pure_output_stationary` (bool):
+    """
+
+    """
+    Explanation of dimensions:
+
+        In case we are handling 6 dimensions and `dim` = 3:
+
+        There are 3 innermost dimensions that are being checked with the template.
+        The other outermost dimensions are not considered.
+
+        The current dimension under foces (d3) is the most outermost dim of the innermost dims.
```

Comment:
I would write as
```suggestion
        The currently focused dimension (d3) is the most outermost dim of the innermost dims.
```
or 
```suggestion
        The current dimension in question (d3) is the most outermost dim of the innermost dims.
```
or
```suggestion
        The current dimension under consideration (d3) is the most outermost dim of the innermost dims.
```